lib/BuildSystem/BuildFile.cpp

```diff
@@ -101,81 +102,105 @@ static void dumpNode(llvm::yaml::Node* node, unsigned indent) {
-        Command *owner = ownerOf(outputNodeAndCommand.first->getName());
-        if (owner == nullptr) {
-          setOwner(outputNodeAndCommand.first, outputNodeAndCommand.second);
-        } else if (owner == outputNodeAndCommand.second) {
-          // A path and some of its subpaths are listed as output dependencies of a task.. Do nothing.
+        if (outputNodeAndCommand.second->isExternalCommand() && outputNodeAndCommand.second->excludeFromOwnershipAnalysis == false) {
+          Command *owner = includedOwnerOf(outputNodeAndCommand.first->getName());
+          if (owner == nullptr) {
+            setOwner(outputNodeAndCommand.first, outputNodeAndCommand.second);
+          } else if (owner == outputNodeAndCommand.second) {
+            // A path and some of its subpaths are listed as output dependencies of a task.. Do nothing.
+          } else {
+            std::vector<Command*> conflictingProducers;
+            conflictingProducers.push_back(outputNodeAndCommand.second);
+            conflictingProducers.push_back(owner);
+            fileDelegate.cannotLoadDueToMultipleProducers(outputNodeAndCommand.first, conflictingProducers);
+            return false;
+          }
         } else {
-          std::vector<Command*> conflictingProducers;
-          conflictingProducers.push_back(outputNodeAndCommand.second);
-          conflictingProducers.push_back(owner);
-          fileDelegate.cannotLoadDueToMultipleProducers(outputNodeAndCommand.first, conflictingProducers);
-          return false;
+          setExcludedOwner(outputNodeAndCommand.first, outputNodeAndCommand.second);
         }
-      }
     }
 
     return true;
   }
 
   /// Check if node is unowned
-  const bool isUnownedNode(const BuildNode* node) {
-    return ownerOf(node->getName()) == nullptr;
+  const bool isIncludedUnownedNode(const BuildNode* node) {
+    return includedOwnerOf(node->getName()) == nullptr && excludedOwnerOf(node->getName()) == nullptr;
   }
 
   /// Set owner
-  void setOwner(BuildNode* node, ExternalCommand* command) {
-    paths[node] = command;
+  void setOwner(BuildNode* node, Command* command) {
+    includedPaths[node] = command;
   }
   
+  /// Set owner of a path that is produced by a command excluded from ownership analysis so we can distinguish it from an unowned path
+  void setExcludedOwner(BuildNode* node, Command* command) {
+    excludedPaths[node] = command;
+  }
+
+  /// Lookup included owner (a directory prefix of inputPath that is included in the analysis)
+  Command* includedOwnerOf(StringRef inputPath) {
+    auto it = std::find_if(includedPaths.begin(), includedPaths.end(), [inputPath](const std::pair<BuildNode*, Command*>& buildNodeAndCommand) -> bool {
+      if (buildNodeAndCommand.first->getName().endswith("/")) {
+        return inputPath.startswith(buildNodeAndCommand.first->getName());
+      } else {
+        return inputPath.startswith(buildNodeAndCommand.first->getName().str() + "/");
```

Comment:
I was running into a case where two non-directory paths were inside `includedPaths` and one was prefix of another one (e.g. `XYZ/xcbuild` and `XYZ/xcbuildservice`).. So when we check for candidate owners, we must ensure that the prefix we check contains a trailing slash.

nit: might be worth it to add a clarifying comment here